formal_spec/nmt.qnt

```diff
@@ -0,0 +1,847 @@
+// -*- mode: Bluespec; -*-
+
+// This file contains a Quint specification of the (part of the) Namespaced Merkle Tree
+// library. 
+// The specification is written in Quint (https://github.com/informalsystems/quint), a modern specification
+// language based on TLA.
+// 
+// The file is split into five modules:
+//  - basics: some basic functions and types
+//  - nmt_helpers: helper function (e.g., generation of trees, pairwise hashing etc.)
+//  - nmt: creation of NMT proof and their verification
+//  - nmtProofVerification: actions that generate proofs and verify them, thereby checking if 
+//    the logic of the spec is sound
+//  - nmtTest: actions that generate proof and non-deterministically corrupt them afterwards,
+//    whose output can be used in test generation
+//  
+//  The specification models Inclusion proofs, but does not model Absence proofs.
+
+module basics{
+
+    type BINARY = List[int]
+
+    def binary(n: int) : BINARY =
+        if (n == 0)
+            [0]
+        else
+            range(0,n).foldl(
+                [],
+                (acc, i) => {
+                    val c = n / 2^i
+                    if (c > 0)
+                        [c % 2].concat(acc)
+                    else
+                        acc
+                }
+            )
+    
+    // this function gets the first power of 
+    // 2 which is greater than `lastNode`
+    def getFullTreeSize(lastNode : int) : int =
+       if (lastNode == 0)
+            1
+        else
+            2^length(binary(lastNode))
+    
+    def sum(list: List[int]) : int =
+        list.foldl(0, (acc, i) => acc + i)
+    
+    def min(S: Set[int]) : int = 
+        // val tentativeMin = chooseSome(S) ---> chooseSome is not supported yet
+        val tentativeMin = -1
+        S.fold(
+            tentativeMin, 
+            (acc, i) => if ((acc == tentativeMin) or (i < acc)) i else acc
+            )
+
+    def listToSet(S: List[int]) : Set[int] = 
+        S.foldl(Set(), (acc, i) => acc.union(Set(i)))
+
+
+    def setToSortedList(S: Set[int]) : List[int] = 
+        S.fold(
+            [], 
+            (acc, i) => acc.concat(
+                [min(S.exclude(listToSet(acc)))]
+                )
+            )
+
+    def max(S: Set[int]) : int = 
+        // val tentativeMax = chooseSome(S) --> chooseSome not supported yet
+        val tentativeMax = -1
+        S.fold(
+            tentativeMax, 
+            (acc, i) => if ((acc == -1) or (i > acc)) i else acc
+            )
+    
+        
+    
+
+    def reverse(L: List[x]) : List[x] =
+        L.foldl([], (acc, i) => [i].concat(acc))
+
+    def getWithDefault(mapX: int -> a, key: int, default: a) : a =
+        if (mapX.keys().contains(key))
+            mapX.get(key)
+        else
+            default        
+
+
+}
+
+module nmt_helpers{
+    import basics.*
+
+    type NODE_IDX = int
+
+    // hashes will be of type WORD
+    type WORD = List[(str, int)]
```

Comment:
Adding some context: the hash function is implemented as a dummy one (identity function) because I was not checking anything about it. I first started with every data being just a string "data". Later, I decided I wanted to also have indicators of the node id because that enables me to just have a look at the the hashes from the proof and know that the right nodes were selected.

To answer now: since hashing is not being modelled, the structure was chosen only for readability in spec-debugging. (For that, a list of only `int`s, without `str`s would do as well.)

Reply:
>  have indicators of the node id 

By id, do you mean the index of the node in the tree? for example, the root is at index 1, its children are indexed as 2 and 3, and so on.